File: java/src/test/java/it/giannibombelli/wsc2026/booking/application/query/BookingQueryServiceTest.java
package it.giannibombelli.wsc2026.booking.application.query;

import it.giannibombelli.wsc2026.booking.domain.booking.Booking;
import it.giannibombelli.wsc2026.booking.domain.booking.BookingId;
import it.giannibombelli.wsc2026.booking.domain.ports.BookingRepository;
import it.giannibombelli.wsc2026.booking.domain.primitive.GiftCardReference;
import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.Description;
import org.junit.jupiter.api.Test;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;

class BookingQueryServiceTest {

    @Test
    void shouldReturnBookingDetailsWhenBookingExists() {
        InMemoryBookingRepository repository = new InMemoryBookingRepository();
        BookingQueryService queryService = new BookingQueryService(repository);
        BookingId bookingId = EntityId.generate(BookingId::new);
        GiftCardReference giftCardReference = new GiftCardReference(UUID.randomUUID());
        Description description = new Description("summer stay");
        Booking booking = Booking.place(bookingId, description, giftCardReference);
        repository.save(booking);

        Optional<BookingDetails> result = queryService.findById(bookingId);

        assertThat(result).isPresent();
        assertThat(result.get().id()).isEqualTo(bookingId.value());
        assertThat(result.get().description()).isEqualTo(description);
        assertThat(result.get().giftCardId()).isEqualTo(giftCardReference.value());
    }

    @Test
    void shouldReturnEmptyWhenBookingDoesNotExist() {
        InMemoryBookingRepository repository = new InMemoryBookingRepository();
        BookingQueryService queryService = new BookingQueryService(repository);
        BookingId bookingId = EntityId.generate(BookingId::new);

        Optional<BookingDetails> result = queryService.findById(bookingId);

        assertThat(result).isEmpty();
    }

    private static final class InMemoryBookingRepository implements BookingRepository {
        private final Map<UUID, Booking> bookings = new HashMap<>();

        @Override
        public void save(Booking booking) {
            bookings.put(booking.id().value(), booking);
        }

        @Override
        public Optional<Booking> findById(BookingId id) {
            return Optional.ofNullable(bookings.get(id.value()));
        }
    }
}

File: typescript/tests/booking/application/query/bookingQueryService.test.ts
import { describe, expect, it } from 'vitest';
import { BookingQueryService } from '@/booking/application/query/bookingQueryService.js';
import { Booking } from '@/booking/domain/booking/booking.js';
import { BookingId } from '@/booking/domain/booking/bookingId.js';
import { BookingRepository } from '@/booking/domain/ports/bookingRepository.js';
import { Description } from '@/common/domain/primitive/description.js';
import { generateId } from '@/common/domain/identity/entityId.js';
import { GiftCardId } from '@/giftcard/domain/giftcard/giftCardId.js';

class InMemoryBookingRepository implements BookingRepository {
    private readonly bookings = new Map<string, Booking>();

    save(booking: Booking): void {
        this.bookings.set(booking.id().value.value, booking);
    }

    findById(id: BookingId): Booking | null {
        return this.bookings.get(id.value.value) ?? null;
    }
}

describe('BookingQueryService', () => {
    it('should return booking details when booking exists', () => {
        const repository = new InMemoryBookingRepository();
        const queryService = new BookingQueryService(repository);
        const bookingId = generateId((value) => new BookingId(value));
        const giftCardId = generateId((value) => new GiftCardId(value));
        const description = new Description('summer stay');
        const booking = Booking.place(bookingId, description, giftCardId);
        repository.save(booking);

        const result = queryService.findById(bookingId);

        expect(result).not.toBeNull();
        expect(result!.id).toBe(bookingId.value.value);
        expect(result!.description).toBe(description);
        expect(result!.giftCardId).toBe(giftCardId.value.value);
    });

    it('should return null when booking does not exist', () => {
        const repository = new InMemoryBookingRepository();
        const queryService = new BookingQueryService(repository);
        const bookingId = generateId((value) => new BookingId(value));

        const result = queryService.findById(bookingId);

        expect(result).toBeNull();
    });
});
